lib/std/os/linux/io_uring.zig

```diff
@@ -0,0 +1,1200 @@
+    defer std.fs.cwd().deleteFile(path) catch {};
+    const fd = file.handle;
+
+    var buffer_write = [_]u8{97} ** 20;
+    var buffer_read = [_]u8{98} ** 20;
+    var sqe_write = try ring.write(123, fd, buffer_write[0..], 10);
+    testing.expectEqual(linux.IORING_OP.WRITE, sqe_write.opcode);
+    testing.expectEqual(@as(u64, 10), sqe_write.off);
+    sqe_write.flags |= linux.IOSQE_IO_LINK;
+    var sqe_read = try ring.read(456, fd, buffer_read[0..], 10);
+    testing.expectEqual(linux.IORING_OP.READ, sqe_read.opcode);
+    testing.expectEqual(@as(u64, 10), sqe_read.off);
+    testing.expectEqual(@as(u32, 2), try ring.submit());
+
+    var cqe_write = try ring.copy_cqe();
+    var cqe_read = try ring.copy_cqe();
+    // Prior to Linux Kernel 5.6 this is the only way to test for read/write support:
+    // https://lwn.net/Articles/809820/
+    if (cqe_write.res == -linux.EINVAL) return error.SkipZigTest;
+    if (cqe_read.res == -linux.EINVAL) return error.SkipZigTest;
+    testing.expectEqual(linux.io_uring_cqe {
+        .user_data = 123,
+        .res = buffer_write.len,
+        .flags = 0,
+    }, cqe_write);
+    testing.expectEqual(linux.io_uring_cqe {
+        .user_data = 456,
+        .res = buffer_read.len,
+        .flags = 0,
+    }, cqe_read);
+    testing.expectEqualSlices(u8, buffer_write[0..], buffer_read[0..]);
+}
+
+test "openat/close" {
+    if (builtin.os.tag != .linux) return error.SkipZigTest;
+
+    var ring = IO_Uring.init(1, 0) catch |err| switch (err) {
+        error.SystemOutdated => return error.SkipZigTest,
+        error.PermissionDenied => return error.SkipZigTest,
+        else => return err
+    };
+    defer ring.deinit();
+
+    const path = "test_io_uring_openat_close";
+    defer std.fs.cwd().deleteFile(path) catch {};
+
+    const flags: u32 = os.O_CLOEXEC | os.O_RDWR | os.O_CREAT;
+    const mode: os.mode_t = 0o666;
+    var sqe_openat = try ring.openat(789, linux.AT_FDCWD, path, flags, mode);
+    testing.expectEqual(io_uring_sqe {
+        .opcode = .OPENAT,
+        .flags = 0,
+        .ioprio = 0,
+        .fd = linux.AT_FDCWD,
+        .off = 0,
+        .addr = @ptrToInt(path),
+        .len = mode,
+        .rw_flags = flags,
+        .user_data = 789,
+        .buf_index = 0,
+        .personality = 0,
+        .splice_fd_in = 0,
+        .__pad2 = [2]u64{ 0, 0 }
+    }, sqe_openat.*);
+    testing.expectEqual(@as(u32, 1), try ring.submit());
+
+    var cqe_openat = try ring.copy_cqe();
+    testing.expectEqual(@as(u64, 789), cqe_openat.user_data);
+    if (cqe_openat.res == -linux.EINVAL) return error.SkipZigTest;
+    // AT_FDCWD is not fully supported before kernel 5.6:
+    // See https://lore.kernel.org/io-uring/20200207155039.12819-1-axboe@kernel.dk/T/
+    // We use IORING_FEAT_RW_CUR_POS to know if we are pre-5.6 since that feature was added in 5.6.
+    if (cqe_openat.res == -linux.EBADF and (ring.features & linux.IORING_FEAT_RW_CUR_POS) == 0) {
+        return error.SkipZigTest;
```

Comment:
We should prefer to only skip at the start of tests rather than in the middle. A skip here suggests this test should be broken into two.

Reply:
You have to make the syscall before you can interrogate the error so you can't skip at the start of the test. Also, `AT_FDCWD` will probably be backported to pre-5.6, so we only want to skip the test because of a potential `EBADF` if the kernel actually gives us an error for `AT_FDCWD`, not because we tried to feature detect upfront.

This test is only doing `openat` and `close`, and the `openat` is the first thing that could fail because of kernel support, so it's not failing in the middle. It just looks that way because we have some init code upfront, but this is actually the first `cqe` we interrogate.

Reply:
On second thought, you're right. If we skip for lack of full `AT_FDCWD` support, then we also skip the `close` test.

These are now split into two tests.